### DISCUSSION TITLE
We need to talk about custom attributes

### DISCUSSION THREAD

**Discussion by genox** [medusajs/medusa#5303] [Q&A]
Hi,

I already sang my praises here and I still think this is a nice tool with a lot of potential but the experience I had in the past few weeks trying to get some simple things going is a bit frustrating and impacts my efficiency massively.

My main pain point are custom attributes. You know, the thing basically every store needs. :) There are some other things but I think this is by far the most important feedback I can give you.

I made my way through extending models, figuring out how to make fields appear in API responses and creating UIs for actually editing them with lots of trial and error. I patched source code files that enable me to import data to populate those fields because that is easier than to rewrite the entire batch job from scratch, duplicating 99% of the code for no reason with all dependencies. Now I moved to the client side and I have to find out that custom attributes have not been considered to be required as filters or searched by.

How are people using custom attributes, at all, without a way to use them in search or filter? Are most of the people using medusa as an order system and manage product and product related content somewhere else? Or does the learning curve involve rewriting every product endpoint from scratch? Or does everybody run an ES instance right from the start?

This feels so much more convoluted, complicated and hard to understand without advanced knowledge of the entire base system than it would need to be. And I think the most important part about all of this is that everybody who wants to use custom attributes has to walk the same rocky road all over again, more or less.

However, I also have to mention that I always received help when I needed it, for which I am immensely thankful.

**Comment by vholik** [comment]
I am using custom attributes with filtering right now on current projects. We are handling them in JSONB metadata.attributes object. Also we filter them by creating custom store/products endpoint and utilizing queryBuilder in service. 

Here is an example of a method in product service: 

```
 const relations = ["variants", "options", "images", "tags"];

    const selects = [
      "product.id",
      "product.title",
      "product.subtitle",
      "product.status",
      "product.description",
      "product.handle",
      "product.thumbnail",
      "product.created_at",
      "product.metadata",
      "prices",
      "category",
      "product.store_id",
      "current_price",
    ];

    const query = this.productRepository_
      .createQueryBuilder("product")
      .leftJoinAndSelect("product.categories", "categories")
      .leftJoinAndSelect("product.variants", "variants")
      .leftJoinAndSelect("variants.prices", "prices")
      .leftJoinAndSelect("variants.options", "options")
      .leftJoinAndSelect("product.images", "images")
      .leftJoinAndSelect("product.tags", "tags")
      .select([...selects, ...relations])
      .where("product.status IN (:...status)", {
        status:  ["published"],
      });
```

And filtering like this: 

```
 if (designer) {
      query.andWhere(
        `product.metadata->'attributes'->>'designer' IN (:...designer)`,
        {
          designer,
        }
      );
    }
```

I heard that Medusa is migrating to MikroORM and that is very awesome because we could filter mikroORM jsonb fields just like a simple field in config. But overall I suggest you to create a model `attribute` and make a relation with products. And you can easily filter `attributes` in config.

**Reply by vholik** [reply]
Update. I made a plugin called `medusa-custom-attributes` go check it out!

**Reply by genox** [reply]
That sounds awesome, thank you for letting me know. I will check it out soon.

**Reply by yf-hk** [reply]
Really, is it true that Medusa is going to use MikroORM? It's more feature rich but it's slower

**Comment by NicolasGorga** [comment]
Hello, reviving this thread. For what i have seen and tried, the medusa-custom-attributes plugin is for defining "markers", to make it easier to filter and search on the Storefront. Apart from having this markers, i also have the need to establish some attributes, which are common to every product or linked to a category (same concept here as defining a custom attribute to a category in the plugin), but for them to impact product_option, so that they can be used to define the product_variants.

Is there anything already developed for this? How are people handling this common use case. I am implementing a second handclothing marketplace, and i want to have attributes like Size and Color, with predefined set of values, as required fields for creating a variant.

**Comment by tylik1** [comment]
medusa-custom-attributes stopped working with v2. No native support for custom attributes in v2, makes me wonder if no one is using it? Also no info in docs on how to overcome this problem (

**Reply by joekendal** [reply]
There is a new way by extending a module, creating a link, using additional_data for api requests then querying with the custom field added.

The docs do explain it but they are very poorly organised

**Comment by OliverRC** [comment]
For me the deal breaker with Medusa thus far as a Magento user is Attributes and AttributeSet.

**Comment by NicolasGorga** [comment]
Hey guys, I've started this plugin https://github.com/NicolasGorga/medusa-product-attributes inspired on custom attributes repo by Viktor, but for V2. So far, it works but i have not created the Admin UI extensions, to manage them from the UI, i manage them via curl.

**Reply by OliverRC** [reply]
Nice, is there plans to do something like attribute sets that can be assigned to a product?
This is the only useful feature of Magento that I've not seen supported on other platforms.

For example say you always want 5 specific attributes for T-Shirts. Then you can assign the attribute set to the product and all the attributes within the set become fields to be filled on that product.

**Reply by NicolasGorga** [reply]
Hey @OliverRC I think that is a really good feature. So far, what is already there and close to what you mention (although simpler, less flexible) is linking attributes to specific categories, so when you link a category to a product, it dinamically updates the attributes you can assign to the product.

Current logic is: category specific attributes + global attributes (those linked to no category). Feel free to open a PR or reach out to me to discuss further, my email is nicogorga11@gmail.com
